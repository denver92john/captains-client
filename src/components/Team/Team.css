.teams-component {
    padding: 20px;
    background-color: ghostwhite;
    min-width: 200px;
    min-height: 200px;
}