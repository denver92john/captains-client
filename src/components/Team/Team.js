import React, {Component} from 'react';
import {Input, Button} from '../Utils/Utils';
import './Team.css';

function createNamesArray(array) {
    const nameArray = [];
    array.map(item => nameArray.push(` ${item.item_name}`))
    return nameArray;
}

function shuffleArray(array) {
    for (let i = array.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [array[i], array[j]] = [array[j], array[i]];
    }
    return array;
}

function createTeams(arr, numberOfTeams) {
    let rest = arr.length % numberOfTeams,
        restUsed = rest,
        partLength = Math.floor(arr.length / numberOfTeams),
        result = [];

    for (let i = 0; i < arr.length; i += partLength) {
        let end = partLength + i,
            add = false;

        if (rest !== 0 && restUsed) {
            end++;
            restUsed--;
            add = true;
        }

        result.push(arr.slice(i, end));

        if (add) {
            i++;
        }
    }
    return result;
}

class Team extends Component {
    state = {
        names: [],
        teams: []
    }

    // do i need to use shouldComponentUpdate()??
    componentDidUpdate(oldProps) {
        if(oldProps.items !== this.props.items) {
            //console.log('yes they diff')
            const listNames = createNamesArray(this.props.items)
            //console.log(listNames)
            this.setState({names: listNames})
        }
    }

    handleSubmit = ev => {
        ev.preventDefault();
        const {number_teams} = ev.target;
        const shuffledList = shuffleArray(this.state.names);
        const newTeams = createTeams(shuffledList, number_teams.value);
        this.setState({teams: newTeams})
    }

    renderForm() {
        return (
            <form onSubmit={this.handleSubmit}>
                    <div className="form-section">
                        <label htmlFor="number-teams-input">Number of teams (2-10): </label>
                        <Input 
                            type="number"
                            id="number-teams-input"
                            name="number_teams"
                            min="2"
                            max="10"
                        />
                    </div>

                    <div className="form-buttons">
                        <Button type="submit">Create Teams</Button>
                    </div>
                </form>
        );
    }

    renderTeamList() {
        const {teams} = this.state;
        return (
            <ul className="named-list">
                {teams.map((team, index) =>
                    <li key={index} className="named-list-item">
                        <p>Team {index + 1}: {team.toString()}</p>
                    </li>
                )}
            </ul>
        );
    }

    render() {
        const {error} = this.props;
        let content;
        if(error) {
            content = <p>There wasn an error</p>
        } else {
            content = 
                <div className="wrapper teams-component">
                    {this.renderForm()}
                    {this.renderTeamList()}
                </div>
        }

        return (
            <>{content}</>
        );
    }
}

export default Team;