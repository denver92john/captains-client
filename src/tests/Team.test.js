import React from 'react';
import ReactDOM from 'react-dom';
import {BrowserRouter} from 'react-router-dom';
import renderer from 'react-test-renderer';
import Team from '../components/Team/Team';

describe('Team component', () => {
    const teamProp = [
        {
            id: 5,
            item_name: "Cormac",
            list_id: 2
        },
        {
            id: 6,
            item_name: "Ken",
            list_id: 2
        }
    ];

    it('renders without errors', () => {
        const div = document.createElement('div');
        ReactDOM.render(
            <BrowserRouter>
                <Team items={teamProp} />
            </BrowserRouter>,
            div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it('renders UI as expected', () => {
        const tree = renderer
            .create(<BrowserRouter>
                <Team items={teamProp} />
            </BrowserRouter>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    })
})