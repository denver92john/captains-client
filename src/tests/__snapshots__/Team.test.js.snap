// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Team component renders UI as expected 1`] = `
<div
  className="wrapper team"
>
  <form
    className="team-form"
    onSubmit={[Function]}
  >
    <div
      className="team-form-section"
    >
      <label
        className="team-form-label"
        htmlFor="number-teams-input"
      >
        Number of teams (2-10): 
      </label>
      <input
        className="team-form-input"
        id="number-teams-input"
        max="10"
        min="2"
        name="number_teams"
        type="number"
      />
    </div>
    <button
      className="team-form-button"
      type="submit"
    >
      Create Teams
    </button>
  </form>
  <ul
    className="team-list"
  />
</div>
`;
